### DISCUSSION TITLE
Some questions about the build system

### DISCUSSION THREAD

**Discussion by ahelwer** [tlaplus/tlapm#141] [Q&A] [answered]
Trying to get oriented in this repo, I have some questions:

1. How does the Dune buildsystem interact with the Make build system in this repo? Is each responsible for different things?
2. `tlapm.opam` provides definitions so TLAPM can be made available as an OPAM package, right?
3. What is the breakdown in responsibilities between `tlapm.opam` and `tlapm.opam.template`?
4. What is the breakdown in responsibilities between `dune` and `dune-project`?

**Comment by kape1395** [comment, marked as answer]
> 1. How does the Dune buildsystem interact with the Make build system in this repo? Is each responsible for different things?

The main build system is `dune`. In the root folder, I have added the Makefile just as a convenience frontend; it delegates most of the tasks to the dune.

At several places, more "script intensive" parts are moved from `dune` to a makefile, e.g. `deps/isabelle/Makefile`, then it is invoked from `deps/isabelle/dune`.

> 2. tlapm.opam provides definitions so TLAPM can be made available as an OPAM package, right?

Yes. But my initial experiments to install tlapm via `opam` failed for some reason. I haven't investigated it further since that's not the current use case. Maybe those problems can be fixed. `tlapm.opam` is generated from `dune-project` mostly.

> 3. What is the breakdown in responsibilities between tlapm.opam and tlapm.opam.template?

`tlapm.opam.template` along with `dune-project` is used to generate `tlapm.opam`.  The essential part that required to use the template is to add post-install action for fixing executable flags `["%{make}%" "-C" "%{lib}%/tlapm" "-f" "Makefile.post-install"]`. See https://dune.readthedocs.io/en/stable/reference/packages.html#generating-opam-files for more details.

> 4. What is the breakdown in responsibilities between dune and dune-project?

`dune-project` is a project-wide description. `dune` files are per-directory.  These files contains different find of instructions (https://dune.readthedocs.io/en/stable/reference/).

**Reply by ahelwer** [reply]
Thanks!
